From 780e3f2e0040349f4f579bb40b1097a1cd542e8b Mon Sep 17 00:00:00 2001
From: Geetha sowjanya <gakula@marvell.com>
Date: Wed, 3 Feb 2021 15:18:32 +0530
Subject: [PATCH 11/11] octeontx2-af: TIM: Set conditional clock always on

commit 4f1b757fd3c91565e2133b586f06cf2371762f93 from
git@git.assembla.com:cavium/WindRiver.linux.git

Due to hardware bug TIM expire counter updates incorrectly
because to wrong clock gating.
This patch adds workaround for this issue by setting
FORCE_CSCLK_ENA=1

Change-Id: I803a44d4cca6a9cf55a1265814ed21d852574aa4
Signed-off-by: Geetha sowjanya <gakula@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/45237
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.h       |  1 +
 drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c | 11 +++++++++++
 drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c   |  3 +++
 3 files changed, 15 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
index e21b00d6575b..6ffa090bfd3c 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
@@ -765,4 +765,5 @@ bool is_parse_nibble_config_valid(struct rvu *rvu,
 				  struct npc_mcam_kex *mcam_kex);
 int rvu_npc_set_parse_mode(struct rvu *rvu, u16 pcifunc, u64 mode, u8 dir,
 			   u64 pkind);
+void rvu_tim_hw_fixes(struct rvu *rvu, int blkaddr);
 #endif /* RVU_H */
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c
index 0c556c3be666..8834edbd735d 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c
@@ -1008,3 +1008,14 @@ void rvu_smqvf_xmit(struct rvu *rvu)
 		otx2smqvf_xmit();
 	}
 }
+
+void rvu_tim_hw_fixes(struct rvu *rvu, int blkaddr)
+{
+	u64 cfg;
+	/* Due wrong clock gating, TIM expire counter is updated wrongly.
+	 * Workaround is to enable force clock (FORCE_CSCLK_ENA = 1).
+	 */
+	cfg = rvu_read64(rvu, blkaddr, TIM_AF_FLAGS_REG);
+	cfg |= BIT_ULL(1);
+	rvu_write64(rvu, blkaddr, TIM_AF_FLAGS_REG, cfg);
+}
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c
index 3c2cda84cf30..ed6143e8c2f4 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c
@@ -337,5 +337,8 @@ int rvu_tim_init(struct rvu *rvu)
 		 BIT_ULL(0); /* ENA_TIM */
 	rvu_write64(rvu, blkaddr, TIM_AF_FLAGS_REG, regval);
 
+	if (is_rvu_otx2(rvu))
+		rvu_tim_hw_fixes(rvu, blkaddr);
+
 	return 0;
 }
-- 
2.26.1

